### DISCUSSION TITLE
Cmake issue

### DISCUSSION THREAD

**Discussion by p-selini** [stephenberry/glaze#1228] [General]
I am trying to use this lib in our project, but when I add the following to my CMake it compalins about in-source build not supported. 
```
FetchContent_Declare(
  glaze
  GIT_REPOSITORY https://github.com/stephenberry/glaze.git
  GIT_TAG v2.9.5
  GIT_SHALLOW TRUE
)

FetchContent_MakeAvailable(glaze)
```
but when I ran cmake I see this error:

**In-source builds are not supported. Please read the BUILDING document before trying to build this project. You may need to delete 'CMakeCache.txt' and 'CMakeFiles/' first.**

A little bit google search shows me to do **cmake -H. -Bbuild**  but it did not help. Any ideas?

**Comment by stephenberry** [comment]
I'd probably have to see more of your CMake to determine what is going on. Are you doing the following?

```
include(FetchContent)

FetchContent_Declare(
  glaze
  GIT_REPOSITORY https://github.com/stephenberry/glaze.git
  GIT_TAG v2.9.5
  GIT_SHALLOW TRUE
)

FetchContent_MakeAvailable(glaze)

add_executable(${PROJECT_NAME} ${srcs})
target_link_libraries(${PROJECT_NAME} PRIVATE glaze::glaze)
```

**Reply by p-selini** [reply]
I have not link to my project yet, just trying to add glaze as dependency. Other than the part I pasted above, I have other packages as dependency which are all fine (examples dependencies, absl, zlib,curl, similar as I quoted in original post), but here are additional non-other dependency entries in cmake file
`include(CheckLibraryExists)`
`include(ExternalProject)`
`include(FetchContent)`
`set(FETCHCONTENT_QUIET OFF)`

here is the full error message when I try to do cmake

> CMake Error at _deps/glaze-src/cmake/prelude.cmake:4 (message):
  In-source builds are not supported.  Please read the BUILDING document
  before trying to build this project.  You may need to delete
  'CMakeCache.txt' and 'CMakeFiles/' first.
Call Stack (most recent call first):
  _deps/glaze-src/CMakeLists.txt:3 (include)

**Reply by stephenberry** [reply]
Does this help? https://stackoverflow.com/questions/45518317/in-source-builds-are-not-allowed-in-cmake

It seems like you just need some sort of build folder.

**Reply by p-selini** [reply]
No, same setting other repo did not ask for this. And by looking at cmake generated file it seems it already created build folder:

> _deps/glaze-build:
> CMakeFiles/
> 
> _deps/glaze-src:
> cmake/  CMakeLists.txt  CMakePresets.json  docs/  include/  LICENSE  README.md  src/  tests/  util/
> 
> _deps/glaze-subbuild:
> CMakeCache.txt  CMakeFiles/  cmake_install.cmake  CMakeLists.txt  glaze-populate-prefix/  Makefile
> 

You can see it created folder `glaze-build` `glaze-subbuild` `glaze-src` 

I took another project in my cmake has same setting but no errors as example:

> _deps/fmt-build:
> CMakeFiles/  cmake_install.cmake  Makefile
> 
> _deps/fmt-src:
> ChangeLog.rst  CMakeLists.txt  CONTRIBUTING.md  doc/  include/  LICENSE.rst  README.rst  src/  support/  test/
> 
> _deps/fmt-subbuild:
> CMakeCache.txt  CMakeFiles/  cmake_install.cmake  CMakeLists.txt  fmt-populate-prefix/  Makefile
> 

you can see you have extra `CMakePresets.json` than `fmt` I am wondering if it is the issue?

**Reply by stephenberry** [reply]
Hmm, I wouldn't think `CMakePresets.json` would be the problem. The in source build error message comes from here: https://github.com/stephenberry/glaze/blob/e0b7ea2beb0d7374f7c3c0ec41d5a980b4cc9fa4/cmake/prelude.cmake#L5

I don't understand why `if(CMAKE_SOURCE_DIR STREQUAL CMAKE_BINARY_DIR)` would be true in your use case.

**Reply by p-selini** [reply]
ah, you solved issue. I think our project (predates me) might set this to equal and dependency lib is within our include I think. So anyway we can disable it through some variable trick (or save it and restore after your project with some sort of post trick)?  I can't change my current project config of those since I did not own it. Also if you do add a variable anyway you can do it for earlier version since we are still c++20 not c++23 yet.

**Reply by stephenberry** [reply]
Okay, that makes sense. I'm glad the issue has been discovered. A lot of Glaze's cmake files are structured to follow https://github.com/friendlyanon/cmake-init, which has nice templates

**Reply by p-selini** [reply]
Like I said I don't control our cmake it is much larger effort, so I can't change theirs. Anyway for you to add a flag to selectively disable this check if I pass that flag on build commandline?

**Reply by stephenberry** [reply]
Yes, I should be able to add an option to disable the check. But, I think it may have larger CMake implications for you, as cmake-init always requires this behavior and is used by a lot of projects. I think it can cause issues with CMake itself, but I can look into it.
